vllm/v1/metrics/perf.py

```diff
@@ -0,0 +1,939 @@
+"""
+Analytic flops/memory estimation module for transformer components.
+"""
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from dataclasses import dataclass
+from typing import Protocol
+
+import torch
+from pydantic import BaseModel, ConfigDict, Field, model_validator
+from typing_extensions import Self
+
+from vllm.config import VllmConfig
+from vllm.logger import init_logger
+from vllm.utils.torch_utils import (
+    STR_DTYPE_TO_TORCH_DTYPE,
+    get_dtype_size,
+    get_kv_cache_torch_dtype,
+)
+from vllm.v1.core.sched.output import SchedulerOutput
+from vllm.v1.metrics.stats import PerfStats
+
+logger = init_logger(__name__)
+
+#### Basic Data Types ####
+
+
+@dataclass(frozen=True)
+class ExecutionContext:
+    """
+    Represents an execution context for a single request.
+
+    FIXME: Might need to extend to batched context later.
+
+    Example)
+    - full prefill with 2048 tokens
+      -> ExecutionContext(2048, 2048, True)
+    - partial prefill with 1024 tokens 2048 context
+      -> ExecutionContext(1024, 2048, True)
+    - decode with 8192 context
+      -> ExecutionContext(1, 8192, False)
+    """
+
+    num_tokens: int
+    context_len: int
+    is_prefill: bool
```

Comment:
Why do we need to distinguish prefill and decode? From flops perspective they should be the same right?

Reply:
For example, the read_bytes() will differ depending on whether it is prefill or decode (i.e. decode will "read" context_len amount of KV cache, whose precision might differ from activation_byte_size, and so on). So I think it's good to keep them separate.